Java-Inheritance.md

```diff
@@ -0,0 +1,172 @@
+
+```java
+public class Vehicle() {
+	public void start() {
+	  System.out.println("Vehicle start code");
+	}
+}
+
+public class Car extends Vehicle {
+	public void start() {
+	  System.out.println("Car start code");
+  }
+}
+
+Car car = new Car();
+car.start(); // "Car start code"
+```
+
+So, it's pretty simple to override methods in the subclass. Although, there is a _catch_. Only that superclass method will be overriden which has the _exact same method signature_ as the subclass method. That means the subclass method definition must have the exact same name, same number and type of parameters, and in the exact same sequence. Thus, `public void start(String key)` would not override `public void start()`.
+
+**Notes** :
+
+* You cannot override private methods of the superclass. (Quite obvious, isn't it?)
+* What if the method of superclass which you are overriding in the subclass suddenly gets obliterated or methods changed? It would fail in runtime! So Java provides you a nifty annotation `@Override` which you can place over the subclass method, which will warn the compiler of those incidents! 
+
+Annotations in Java is a good coding practice, but they are not a necessity. The compiler is smart enough to figure out overriding on its own though. Unlike other OOP languages, Annotations in Java it doesn't necessarily modify the method or add extra functionality. Read more about [Java Annotations]() here.
+
+## How to call super class methods?
+
+Funny you ask about it! Just use the keyword `super` :
+
+```java
+public class Vehicle() {
+	public void start() {
+	  System.out.println("Vehicle start code");
+	}
+}
+
+public class Car extends Vehicle {
+	public void run() {
+	  super.start();
+  }
+}
+
+Car car = new Car();
+car.run(); // "Vehicle start code"
+```
+
+**N.B.** : Although you can call the parent method by `super` call, you cannot go up the inheritance with chained `super` calls.
+
+## How to know the type of a class?
+
+Using the `instanceof` keyword. Having lots of classes and subclasses it would be a little confusing to know which class is a subclass of which one in runtime. So, we can use `instanceof` to determine whether a class is actually a subclass of a superclass or not.
+
+```java
+Car car = new Car();
+
+boolean flag = car instanceof Vehicle; // true in this case!
+```
+
+## Constructors & Inheritance
+
+As mentioned earlier, constructors cannot be directly inherited by a subclass. Although, a subclass is _required_ to call its parent's constructor as the [first thing](http://stackoverflow.com/questions/1168345/why-does-this-and-super-have-to-be-the-first-statement-in-a-constructor) in its own constructor. How? You guessed it, using `super` :
+
+```java
+public class Vehicle() {
+	public Vehicle() {
+		// constructor
+	}
+	public void start() {
+	  System.out.println("Vehicle start code");
+	}
+}
+
+public class Car extends Vehicle {
+	public Car() {
+		super();
```

Comment:
Invocation to `super` constructor is done in the case when the super class is to be called with any other constructor other than the default constructor.